### PR TITLE
use engine flox for ordered groups

### LINKED ISSUE #233: optimize groupby for resample
I keep seeing this pattern at NCAR

```
MOC.groupby('time.year').mean('time').rename({'year':'time'})
```

This is really a `MOC.resample(time="A").mean()` but with a more desirable time coordinate.

So there's some value in detecting group patterns

### LINKED ISSUE #262: use engine flox if array is ordered?
I find it a bit unfortunate that a quite common case (ordered groups) is 5-15 times slower with flox installed than without (pydata/xarray#7730). One option is to prefer `engine="flox"` if `group_idx` is ordered. 

Is that something that could be considered? Or would that cause problem for the duck array support as discussed in [engines](https://flox.readthedocs.io/en/latest/engines.html)?

I thought about opening an issue in numpy-groupies to add `reduceat` but maybe it's more relevant here.

### PR DESCRIPTION
**Pull request by mathause**
- Closes #233
- Closes #262
- Closes https://github.com/pydata/xarray/issues/7730

Set `engine=None` and select `engine="flox"` for ordered groups. Some comments

- this is certainly a very rough WIP
- I think there could be other logic to determine the optimal engine
- I only determine the engine in `groupby_reduce` however it is already accessed in `xarray_reduce` - this is not optimal (see below)
- the testing approach is brute force at the moment - I think we'd need to test if the correct engine is choosen (and not just run all the tests)
- so the engine detection should be done in a separate function and then we can test this function
- I tested it and this would solve https://github.com/pydata/xarray/issues/7730 (at least for ordered groupings)

https://github.com/xarray-contrib/flox/blob/5ea713efb5019ff9627893060a6c707a1d14170a/flox/xarray.py#L367

### CODE REVIEW

File: flox/core.py [hunk 1]
```diff
@@ -1755,7 +1755,7 @@ def groupby_reduce(
     dtype: np.typing.DTypeLike = None,
     min_count: int | None = None,
     method: T_Method = "map-reduce",
-    engine: T_Engine = "numpy",
+    engine: T_Engine = None,
```
Comment (dcherian) [inline flox/core.py:1758]: Needs a docstring update.

File: flox/core.py [hunk 2]
```diff
@@ -1851,17 +1851,27 @@ def groupby_reduce(
...
+    by_is_dask = tuple(is_duck_dask_array(b) for b in bys)
+    any_by_dask = any(by_is_dask)
+
+    if engine is None:
+        # choose numpy per default
+        engine = "numpy"
+
+        if nby == 1 and not any_by_dask and bys[0].ndim == 1:
+            # maybe move to helper function
+            issorted = lambda arr: (arr[:-1] <= arr[1:]).all()
```
Comment (dcherian) [inline flox/core.py:1865]: yes helper funciton here would be nice.

File: flox/core.py [hunk 3]
```diff
@@ -1851,17 +1851,27 @@ def groupby_reduce(
     xarray.xarray_reduce
     """
 
+    bys: T_Bys = tuple(np.asarray(b) if not is_duck_array(b) else b for b in by)
+    nby = len(bys)
+    by_is_dask = tuple(is_duck_dask_array(b) for b in bys)
+    any_by_dask = any(by_is_dask)
+
+    if engine is None:
```
Comment (dcherian) [inline flox/core.py:1859]: let's make this a helper function with some nice unit tests:
- sorted `by` -> flox
- unsorted `by` -> numpy
- `by` is dask.
- nD `by`
- `nby` == `
- `argmax` does not give engine="flox"

File: flox/xarray.py [hunk 4]
```diff
@@ -72,7 +72,7 @@ def xarray_reduce(
     fill_value=None,
     dtype: np.typing.DTypeLike = None,
     method: str = "map-reduce",
-    engine: str = "numpy",
+    engine: str = None,
```
Comment (dcherian) [inline flox/xarray.py:75]: Again needs docstring update.

File: tests/conftest.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 import pytest
 
 
-@pytest.fixture(scope="module", params=["flox", "numpy", "numba"])
+@pytest.fixture(scope="module", params=[None, "flox", "numpy", "numba"])
```
Comment (dcherian) [inline tests/conftest.py:4]: Lets put this in the benchmarks too.

**Review by dcherian — COMMENTED** [review]
Thanks for working on this!

> so the engine detection should be done in a separate function and then we can test this function

Yes!

> I only determine the engine in groupby_reduce however it is already accessed in xarray_reduce - this is not optimal (see below)

Yes a helper function will fix this. `xarray_reduce` can jsut call that.

**Comment by mathause** [comment]
Thanks for the feedback. I have to let this rest until I find some time again - so feel free to take this over if anyone is interested.

**Comment by mathause** [comment]
I hardly qualify to be on here 😅 thanks for finishing up - this will be a huge win!

**Pull request merged**